compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -96,20 +96,66 @@ bool isBatchMatmulEncodingUser(EncodingUser user) {
   return user == EncodingUser::BATCH_MATMUL;
 }
 
-MaterializeEncodingInfo getEncodingInfoForMatmul(EncodingUser user,
-                                                 EncodingRole role,
+bool getIntOrZero(IntegerAttr a) { return a == IntegerAttr() ? 0 : a.getInt(); }
+
+bool isVecmatEncoding(EncodingAttr encoding) {
+  return encoding.getUser().getValue() == EncodingUser::MATMUL &&
+         getIntOrZero(encoding.getMatmulNarrow_M()) == 1;
+}
+
+bool isMatvecEncoding(EncodingAttr encoding) {
+  return encoding.getUser().getValue() == EncodingUser::MATMUL &&
+         getIntOrZero(encoding.getMatmulNarrow_N()) == 1;
+}
+
+bool isBatchVecmatEncoding(EncodingAttr encoding) {
+  return encoding.getUser().getValue() == EncodingUser::BATCH_MATMUL &&
+         getIntOrZero(encoding.getMatmulNarrow_M()) == 1;
+}
+
+bool isBatchMatvecEncoding(EncodingAttr encoding) {
+  return encoding.getUser().getValue() == EncodingUser::BATCH_MATMUL &&
+         getIntOrZero(encoding.getMatmulNarrow_N()) == 1;
+}
+
+bool isVectorEncoding(int64_t rank, EncodingUser user) {
+  return rank == 1 || (isBatchMatmulEncodingUser(user) && rank == 2);
+}
+
+MaterializeEncodingInfo getEncodingInfoForMatmul(EncodingAttr encoding,
+                                                 int64_t rank,
                                                  TileMxNxK tileMxNxK) {
+  EncodingUser user = encoding.getUser().getValue();
+  EncodingRole role = encoding.getRole().getValue();
+  bool isVector = isVectorEncoding(rank, user);
+  bool isVecmatVector = (isVector && (isVecmatEncoding(encoding) ||
```

Comment:
I'm curious if we can simplify this further. It's OK the way it currently is, but if there's a chance to avoid relying on `matmul_narrow_M` here, that's interesting.

Can this be simplified to the following?

```
bool isVecmatVector = isVector && role == EncodingRole::LHS;
bool isMatvecVector = isVector && role == EncodingRole::RHS;
```

Reply:
I think we are circling back around to the same issue as before -- this works for everything but the RESULT cases. The reason the check exists as it does is that we can correctly direct the result into the vecmat/matvec based on the narrowness.